compiler-core/src/language_server/type_annotations.rs

```diff
@@ -0,0 +1,92 @@
+use super::code_action::CodeActionBuilder;
+use crate::{
+    ast::{TypedFunction, TypedModuleConstant},
+    line_numbers::LineNumbers,
+    type_::pretty::Printer,
+};
+use lsp_types::{
+    CodeActionKind, InlayHint, InlayHintKind, InlayHintLabel, Position, Range, TextEdit, Url,
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct TypeAnnotations {
+    annotations: Vec<(Position, String)>,
+}
+
+impl TypeAnnotations {
+    pub fn from_module_constant(
+        constant: &TypedModuleConstant,
+        line_numbers: &LineNumbers,
+    ) -> Self {
+        if constant.annotation.is_none() {
+            let linecol = line_numbers.line_and_column_number(constant.location.end);
+            let position = Position::new(linecol.line - 1, linecol.column - 1);
+            let type_text = Printer::new().pretty_print(&constant.type_, 0);
+            let text = format!(": {type_text}");
+
+            Self {
+                annotations: vec![(position, text)],
+            }
+        } else {
+            Self::default()
+        }
+    }
+
+    pub fn from_function_definition(function: &TypedFunction, line_numbers: &LineNumbers) -> Self {
+        let mut annotations = vec![];
+        if function.return_annotation.is_none() {
+            let linecol = line_numbers.line_and_column_number(function.location.end);
+            let position = Position::new(linecol.line - 1, linecol.column - 1);
+            let type_text = Printer::new().pretty_print(&function.return_type, 0);
+            let text = format!(" -> {type_text}");
+            annotations.push((position, text))
+        }
+        for argument in &function.arguments {
+            if argument.annotation.is_none() {
+                let linecol = line_numbers.line_and_column_number(argument.location.end);
+                let position = Position::new(linecol.line - 1, linecol.column - 1);
+                let type_text = Printer::new().pretty_print(&argument.type_, 0);
+                let text = format!(": {type_text}");
+                annotations.push((position, text));
+            }
+        }
+
+        Self { annotations }
+    }
+
+    pub fn into_inlay_hints(self) -> impl Iterator<Item = InlayHint> {
+        self.annotations
+            .into_iter()
+            .map(|(position, text)| InlayHint {
+                position,
+                kind: Some(InlayHintKind::TYPE),
+                tooltip: None,
+                padding_left: text.starts_with(' ').then_some(true),
+                padding_right: None,
+                data: None,
+                label: InlayHintLabel::String(text.trim_start().to_owned()),
+                text_edits: Some(vec![TextEdit::new(Range::new(position, position), text)]),
+            })
+    }
+
+    pub fn into_code_action(self, uri: Url) -> Option<CodeActionBuilder> {
```

Comment:
This isn't always used so let's take it by reference, rather than force the clone.
```suggestion
    pub fn into_code_action(self, uri: &Url) -> Option<CodeActionBuilder> {
```